pkg/registry/query.go

```diff
@@ -0,0 +1,322 @@
+					//   the sqlite server and returns seemingly invalid channel entries.
+					//      Don't worry about this. Not used anymore.
+
+					entries = append(entries, channelEntriesForBundle(*b, true)...)
+				}
+			}
+		}
+	}
+	if len(entries) == 0 {
+		return nil, fmt.Errorf("no channel entries found that provide group:%q version:%q kind:%q", group, version, kind)
+	}
+	return entries, nil
+}
+
+// TODO(joelanford): Need to review the expected functionality of this function. I ran
+//   some experiments with the sqlite version of this function and it seems to only return
+//   channel heads that provide the GVK (rather than searching down the graph if parent bundles
+//   don't provide the API). Based on that, this function currently looks at channel heads only.
+//   ---
+//   Separate, but possibly related, I noticed there are several channels in the channel entry
+//   table who's minimum depth is 1. What causes 1 to be minimum depth in some cases and 0 in others?
+func (q Querier) GetLatestChannelEntriesThatProvide(_ context.Context, group, version, kind string) ([]*ChannelEntry, error) {
+	var entries []*ChannelEntry
+
+	for _, pkg := range q.pkgs {
+		for _, ch := range pkg.Channels {
+			b, err := ch.Head()
+			if err != nil {
+				return nil, fmt.Errorf("package %q, channel %q has invalid head: %v", pkg.Name, ch.Name, err)
+			}
+
+			provides, err := doesModelBundleProvide(*b, group, version, kind)
+			if err != nil {
+				return nil, err
+			}
+			if provides {
+				entries = append(entries, channelEntriesForBundle(*b, false)...)
+			}
+		}
+	}
+	if len(entries) == 0 {
+		return nil, fmt.Errorf("no channel entries found that provide group:%q version:%q kind:%q", group, version, kind)
+	}
+	return entries, nil
+}
+
+func (q Querier) GetBundleThatProvides(ctx context.Context, group, version, kind string) (*api.Bundle, error) {
+	latestEntries, err := q.GetLatestChannelEntriesThatProvide(ctx, group, version, kind)
+	if err != nil {
+		return nil, err
+	}
+
+	// It's possible for multiple packages to provide an API, but this function is forced to choose one.
+	// To do that deterministically, we'll pick the the bundle based on a lexicographical sort of its
+	// package name.
+	sort.Slice(latestEntries, func(i, j int) bool {
+		return latestEntries[i].PackageName < latestEntries[j].PackageName
+	})
+
+	for _, entry := range latestEntries {
+		pkg, ok := q.pkgs[entry.PackageName]
+		if !ok {
+			// This should never happen because the latest entries were
+			// collected based on iterating over the packages in q.pkgs.
+			continue
+		}
+		if entry.ChannelName == pkg.DefaultChannel.Name {
+			return q.GetBundle(ctx, entry.PackageName, entry.ChannelName, entry.BundleName)
+		}
+	}
+	return nil, fmt.Errorf("no entry found that provides group:%q version:%q kind:%q", group, version, kind)
+}
+
+func doesModelBundleProvide(b model.Bundle, group, version, kind string) (bool, error) {
```

Comment:
I wonder if it is beneficial to convert this func and `bundleReplaces` to public funcs on `model` packages.
For example:
`func(b model.Bundle) Replaces(name string) bool {}`
Just a thought.

Reply:
At one point, I think I had a version of the model package that had these kind of functions. Let me see if I can work out why I changed it.

Reply:
Ahh. I think I did it this way because I wanted to keep property the model out of the business of parsing and interpreting properties, which would be necessary to answer these types of questions.

Reply:
We _could_ put a `DoesReplace` function in model since `model.Bundle` has a single `Replaces` field in the struct. But the provided GVKs are all stored in opaque properties in the model.

My instinct is to leave the model pretty minimal. If there's a future where multiple packages have to implement these same sort of functions against the model, we could revisit.

Reply:
I see. I agree that this func `doesModelBundleProvide` isn't necessarily beneficial to have on the model because it does require some conversion. The replace one is more straightforward to have on model. Either way, it is not a big deal. I just thought about reorganizing the codebase a bit but it is not a concern at the moment. Move on.